parlai/core/torch_agent.py

```diff
@@ -0,0 +1,260 @@
+
+        xs = padded_xs
+
+        eval_labels_avail = any(['eval_labels_vec' in ex for ex in exs])
+        labels_avail = any(['labels_vec' in ex for ex in exs])
+        some_labels_avail = eval_labels_avail or labels_avail
+
+        # set up the target tensors
+        ys = None
+        labels = None
+        if some_labels_avail:
+            # randomly select one of the labels to update on (if multiple)
+            if labels_avail:
+                num_choices = [len(ex.get('labels_vec', [])) for ex in exs]
+                choices = [random.choice(range(num)) if num != 0 else -1
+                           for num in num_choices]
+                label_vecs = [ex['labels_vec'][choices[i]]
+                              if choices[i] != -1 else torch.LongTensor([])
+                              for i, ex in enumerate(exs)]
+                labels = [ex['labels'][choices[i]]
+                          if choices[i] != -1 else ''
+                          for i, ex in enumerate(exs)]
+            else:
+                num_choices = [len(ex.get('eval_labels_vec', [])) for ex in exs]
+                choices = [random.choice(range(num)) if num != 0 else -1
+                           for num in num_choices]
+                label_vecs = [ex['eval_labels_vec'][choices[i]]
+                              if choices[i] != -1 else torch.LongTensor([])
+                              for i, ex in enumerate(exs)]
+                labels = [ex['eval_labels'][choices[i]]
+                          if choices[i] != -1 else ''
+                          for i, ex in enumerate(exs)]
+            y_lens = [y.shape[0] for y in label_vecs]
+            padded_ys = torch.LongTensor(len(exs),
+                                         max(y_lens)).fill_(self.NULL_IDX)
+            if self.use_cuda:
+                padded_ys = padded_ys.cuda()
+            for i, y in enumerate(label_vecs):
+                if y.shape[0] != 0:
+                    padded_ys[i, :y.shape[0]] = y
+            ys = padded_ys
+
+        return Batch(xs, ys, labels, valid_inds)
+
+    def unmap_valid(self, predictions, valid_inds, batch_size):
+        """Re-order permuted predictions to the initial ordering, includes the
+        empty observations.
+
+        :param predictions: output of module's predict function
+        :param valid_inds: original indices of the predictions
+        :param batch_size: overall original size of batch
+        """
+        unpermuted = [None]*batch_size
+        for pred, idx in zip(predictions, valid_inds):
+            unpermuted[idx] = pred
+        return unpermuted
+
+    def maintain_dialog_history(self, observation, reply='',
+                                useStartEndIndices=True,
+                                splitSentences=False):
+        """Keeps track of dialog history, up to a truncation length.
+        Either includes replies from the labels, model, or not all using param 'replies'."""
```

Comment:
actually, can you update this function to use the current maintain_dialog_history? made a couple changes in another diff to the useStartEndIndices functionality (which should default to False)